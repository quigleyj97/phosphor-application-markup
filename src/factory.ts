import { AttachedProperty } from "@phosphor/properties";
import { Widget } from "@phosphor/widgets";

/**
 * A widget factory for PAM applications.
 */
export class WidgetFactory {
    private readonly ctorRegistry = new Map<string, WidgetFactory.Constructor>();
    private readonly propRegistry = new Map<string, WidgetFactory.IAttachedProperty>();

    /**
     * Register a new widget into the factory.
     *
     * @param name What name this widget will be referenced by in the markup.
     * @param ctor The constructor for the widget being registered.
     * 
     * #### Notes
     *
     * The constructor should take no arguments.
     *
     * TODO: Should we add additional props for things like attribute validation
     * and has-children?
     * 
     * TODO: Should we perform normalization? HTML (and SGML in general) is
     * generally case insensitive, whereas XML is generally case sensitive. This
     * may cause problems, either in user understanding or interfacing with the
     * HTML DOM.
     */
    public registerWidget(name: string, ctor: WidgetFactory.Constructor) {
        if (this.ctorRegistry.has(name)) {
            throw Error("Widget " + name + " is already registered");
        }
        this.ctorRegistry.set(name, ctor);
    }

    /**
     * Register an attached property into the factory.
     * 
     * @param name What name this property will be referenced by.
     * @param prop The AttachedProperty being registered.
     * 
     * #### Notes
     * 
     * Attached property names **must** begin with an uppercase letter! This
     * allows the loader to differentiate between HTML attributes and attached
     * properties.
     * They must also be valid XML attribute names, but this function won't
     * check for that. Of note; attribute names _may_ include a dot.
     */
    public registerProperty(
        name: string,
        prop: WidgetFactory.IAttachedProperty)
    {
        if (this.propRegistry.has(name)) {
            throw Error("Property " + name + " is already registered");
        }
        // TODO: Is there a better way to check for upper-case-ness?
        if (name[0].toLocaleUpperCase() !== name[0]) {
            throw Error("Property name must begin with an uppercase letter");
        }
        this.propRegistry.set(name, prop);
    }

    /**
     * Instantiate a new widget from the factory.
     *
     * @param name The registered name of the widget to create
     */
    public createWidget(name: string): Widget {
        if (!this.ctorRegistry.has(name)) {
            throw Error("Widget " + name + " not recognized!");
        }
        return new (this.ctorRegistry.get(name)!)();
    }

    /**
     * Get an attached property from the factory.
     * 
     * @param name The registered name of the property
     */
    public getProperty(name: string): WidgetFactory.IAttachedProperty {
        if (!this.propRegistry.has(name)) {
            throw Error("Property " + name + " not recognized!");
        }
        return this.propRegistry.get(name)!;
    }
}
export namespace WidgetFactory {
    /** A default global factory for PAM */
    export const global = new WidgetFactory();

    /** A restricted interface for attached properties.
     *
     * This is necessary to work around the fact that many Phosphor framework
     * properties are not directly exposed- things like BoxLayout.Stretch are
     * implemented using this restricted interface.
    */
    export interface IAttachedProperty<T = unknown, U = unknown> {
        get(owner: T): U;
        set(owner: T, value: U): void; 
    }

    /** A convenience type to express a constructor in a covariant way */
    export type Constructor = {
        new(...args: any[]): Widget
    }
}
