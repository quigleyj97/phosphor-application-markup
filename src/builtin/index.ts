import { BoxPanel, Panel, SplitPanel, StackedPanel, TabPanel, Widget, BoxLayout, Layout, SplitLayout } from "@phosphor/widgets";
import { DOMWrapper } from "./wrapper";
import { WidgetFactory } from "../factory";

/**
 * Register PAM default items to a factory.
 * 
 * This is called automatically for the global factory, but will need to be
 * manually called for anything else.
 *
 * @param factory The factory to register against
 */
export function RegisterDefaults(factory: WidgetFactory) {
    //#region DOM wrappers
    factory.registerWidget("Image", DOMWrapper("img"));
    factory.registerWidget("Label", DOMWrapper("label"));
    //#endregion

    //#region Phosphor panels and attached properties
    factory.registerProperty("Layout.HorizontalAlignment", {
        get: (owner) => Layout.getHorizontalAlignment(owner),
        set: (owner, value) => Layout.setHorizontalAlignment(owner, value as Layout.HorizontalAlignment)
    });
    factory.registerProperty("Layout.VerticalAlignment", {
        get: (owner) => Layout.getVerticalAlignment(owner),
        set: (owner, value) => Layout.setVerticalAlignment(owner, value as Layout.VerticalAlignment)
    });
    factory.registerWidget("BoxPanel", BoxPanel);
    factory.registerProperty("BoxLayout.SizeBasis", {
        get: (owner) => ""+BoxLayout.getSizeBasis(owner),
        set: (owner, value) => BoxLayout.setSizeBasis(owner, +value)
    });
    factory.registerProperty("BoxLayout.Stretch", {
        get: (owner) => ""+BoxLayout.getStretch(owner),
        set: (owner, value) => BoxLayout.setStretch(owner, +value)
    });
    factory.registerWidget("Panel", Panel);
    factory.registerWidget("SplitPanel", SplitPanel);
    factory.registerProperty("SplitLayout.Stretch", {
        get: (owner) => ""+SplitLayout.getStretch(owner),
        set: (owner, value) => SplitLayout.setStretch(owner, +value)
    });
    factory.registerWidget("StackedPanel", StackedPanel);
    factory.registerWidget("TabPanel", TabPanel);
    //#endregion

    // TODO: A set of more useful defaults
}
