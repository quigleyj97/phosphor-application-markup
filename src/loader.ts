import { WidgetFactory } from "./factory";
import { MessageLoop } from "@phosphor/messaging";
import { Widget } from "@phosphor/widgets";

/**
 * Injests some markup, and returns a widget
 */
export class MarkupLoader {
    private factory: WidgetFactory;

    constructor(factory = WidgetFactory.global) {
        this.factory = factory;
    }

    /**
     * Load some PAML as XML, returning a valid Phosphor widget.
     * 
     * ## Notes
     * 
     * The loader walks through the XML tree, breadth-first, to create a widget
     * matching the markup. For each element it encounters, it creates a
     * corresponding widget using the WidgetFactory, and then sets any attached
     * properties by looking for attributes that begin with an upper-case
     * letter.
     * 
     * Node attributes that begin with a lowercase letter are correspondingly
     * set on the DOM, including data-attrs.
     *
     * ### Caveats
     *
     * Phosphor attached properties are hacked in right now, as most of them
     * aren't publicly exposed.
     * 
     */
    public loadXml(input: string) {
        const reader = new DOMParser();
        let doc: Document;
        try {
            doc = reader.parseFromString(input, "application/xml");
        }
        catch (err) {
            console.error("[PAM- Loader]", "Failed to load XML- could not parse markup");
            throw err; // nothing better to do
        }
        const errorDoc = doc.getElementsByTagName("parsererror");
        if (errorDoc.length > 0) {
            // There's an error in the document, but the user agent was either
            // able to partially parse it or opted not to throw an error.
            console.error("[PAM- Loader]", "Failed to load XML- markup error");
            console.error(errorDoc);
            throw new Error("PAM: ParserError");
        }
        // Return the root element.
        return this.instantiateWidgetForNode(doc.documentElement);
    }

    private instantiateWidgetForNode(node: Element): Widget {
        const name = node.tagName;
        const widget = this.factory.createWidget(name);
        for (const attr of node.getAttributeNames()) {
            // assumption: non-null, since it was returned by getAttributeNames 
            const attrValue = node.getAttribute(attr)!;

            // if the attribute name begins with an uppercase letter, it's an
            // attached property
            if (attr[0].toLocaleUpperCase() === attr[0]) {
                const property = this.factory.getProperty(attr);
                // TODO: JS types need to be figured out.
                property.set(widget, attrValue);
            } else {
                widget.node.setAttribute(attr, attrValue);
            }
        }
        // This iterates through _all_ nodes, including text and comment nodes
        for (const childNode of node.childNodes) {
            // If it's a text node or a comment node, insert it directly into
            // the DOM.
            if (childNode instanceof Text || childNode instanceof Comment) {
                widget.node.appendChild(childNode);
                continue;
            }
            if (!(childNode instanceof Element)) {
                // TODO: Unsupported node- it's neither a comment nor text, but
                // it isn't an element either.
                // For security reasons, it's best not to copy unfamiliar things
                // though since this is an alpha, log it to the console for
                // later inspection
                console.warn("[PAM- Loader]", "Unsupported element enountered");
                console.log(childNode);
                console.log("This node will not be copied to HTML");
                continue;
            }
            // else it's a full node, process it
            try {
                const childWidget = this.instantiateWidgetForNode(childNode);
                // Message the widget about this new child
                const msg = new Widget.ChildMessage("child-added", childWidget);
                MessageLoop.postMessage(widget, msg);
            } catch (err) {
                // continue, but warn.
                console.warn("[PAM- Loader]", "Failed to create child node");
                console.warn(err);
            }
        }
        return widget;
    }
}
