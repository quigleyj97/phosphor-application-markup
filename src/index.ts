import { WidgetFactory } from "./factory";
import { RegisterDefaults } from "./builtin";
import { MarkupLoader } from "./loader";

RegisterDefaults(WidgetFactory.global);

export {
    WidgetFactory,
    MarkupLoader,
    RegisterDefaults
};
