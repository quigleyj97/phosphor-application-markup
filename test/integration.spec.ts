import { readFileSync } from "fs";
import { MarkupLoader } from "../src";
import { MessageLoop } from "@phosphor/messaging";

describe("Integration test", () => {
    let file: string;

    beforeAll(() => {
        file = readFileSync("./examples/testdoc.xml", {encoding: "utf8"});
    })

    test("Should render the HelloWorld example", () => {
        let loader = new MarkupLoader();
        const out = loader.loadXml(file);
        MessageLoop.flush();
        expect(out.node.outerHTML).toMatchSnapshot();
    })
})