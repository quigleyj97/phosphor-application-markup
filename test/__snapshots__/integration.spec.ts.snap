// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test Should render the HelloWorld example 1`] = `
"<div class=\\"p-Widget p-Panel p-BoxPanel\\" data-direction=\\"top-to-bottom\\" data-alignment=\\"start\\">
    <label class=\\"p-Widget p-BoxPanel-child\\" style=\\"position: absolute;\\">
        Hello, world!
    </label>
    <div class=\\"p-Widget p-Panel p-BoxPanel p-BoxPanel-child\\" data-direction=\\"left-to-right\\" data-alignment=\\"start\\" style=\\"position: absolute;\\">
        <img class=\\"p-Widget p-BoxPanel-child\\" href=\\"https://placekitten.com/g/200/300\\" alt=\\"Kitten\\" style=\\"position: absolute;\\">
        <img class=\\"p-Widget p-BoxPanel-child\\" href=\\"http://place-puppy.com/200x200\\" alt=\\"Pupper\\" style=\\"position: absolute;\\">
    </div>
</div>"
`;
